### DISCUSSION TITLE
Use private registry running on host machine

### DISCUSSION THREAD

**Discussion by SanjayVas** [k3s-io/k3s#7149] [Q&A] [answered]
I'm trying to swap over from KinD to K3s as the latter is much lighter weight. The main issue I'm having is how to get K3s to use a private registry that's running on my host machine. I'm using rootless mode in case that makes a difference.

I followed [the instructions](https://docs.k3s.io/installation/private-registry) for configuring the private registry, using `~/.rancher/k3s/etc/rancher/k3s` instead of `/etc/rancher/k3s`. This didn't end up resulting in the containerd `config.toml` file getting updated. I found I had to instead specify `--private-registry` in the `k3s server` command in the systemd unit. Doing this then resulted in the `config.toml` file having the right configuration. I'm still seeing `ErrImagePull` with "connect: connection refused". I think the issue is that the registry is running on the K3s host, meaning it's at `localhost:5000`.

repositories.yaml:
```yaml
mirrors:
  "localhost:5000":
    endpoint:
    - "http://localhost:5000"
```

This is very similar to the [configuration for KinD](https://kind.sigs.k8s.io/docs/user/local-registry/), though that has an extra step of connecting the Docker network for KinD to the registry. Is there something similar that can be done with K3s? I'd rather not just expose the entire host network.

**Comment by brandond** [comment, marked as answer]
Have you tried using the host's actual name or address, instead of localhost? What addresses is the registry service listening on? Rootless mode runs in a separate network namespace, so localhost in the child namespace is not the same as it is if it was running normally on the host. It's like halfway to running it in Docker, which might actually be easier if you're coming from kind.

**Reply by SanjayVas** [reply]
It was bound to the loopback address. I bound it to the other interface IP and updated the mirrors endpoint. That worked, but I'm a bit concerned as it means I'm having the registry accessible to my entire network.

**Reply by SanjayVas** [reply]
I tried going back to the loopback address 127.0.0.1 and setting `Environment=K3S_ROOTLESS_DISABLE_HOST_LOOPBACK=false` in my systemd unit file. Unfortunately that did not work. It seems that the only solution is to use a non-loopback address.

**Reply by SanjayVas** [reply]
I got it working with loopback using an IP alias (`sudo ifconfig lo:0 192.168.50.1`)
